### PR TITLE
solution 29/09

### PR DESCRIPTION
**Pull request by VadimKadasiev**

### CODE REVIEW

File: src/main/java/core/basesyntax/ColorSupplier.java [hunk 1]
```diff
@@ -0,0 +1,20 @@
+package core.basesyntax;
+
+import java.util.Random;
+
+public class ColorSupplier {
+    private Random random = new Random();
+    private int index = getRandom().nextInt(Colors.values().length);
```
Comment (kozhukhovsky) [inline src/main/java/core/basesyntax/ColorSupplier.java:7]: So there is no accident. When initializing a ColorSupplier object, the index is created once and will always be used only.

File: src/main/java/core/basesyntax/ColorSupplier.java [hunk 2]
```diff
@@ -0,0 +1,20 @@
...
+public class ColorSupplier {
+    private Random random = new Random();
+    private int index = getRandom().nextInt(Colors.values().length);
+
+    public int getIndex() {
+        return index;
+    }
+
+    public String getRandomColor() {
+        return "" + Colors.values()[getIndex()];
```
Comment (kozhukhovsky) [inline src/main/java/core/basesyntax/ColorSupplier.java:14]:
```suggestion
        int index = getRandom().nextInt(Colors.values().length);
        return Colors.values()[index].name();

```

File: src/main/java/core/basesyntax/ColorSupplier.java [hunk 3]
```diff
@@ -0,0 +1,20 @@
...
+
+import java.util.Random;
+
+public class ColorSupplier {
+    private Random random = new Random();
+    private int index = getRandom().nextInt(Colors.values().length);
+
+    public int getIndex() {
+        return index;
+    }
```
Comment (kozhukhovsky) [inline src/main/java/core/basesyntax/ColorSupplier.java:11]:
```suggestion

```

File: src/main/java/core/basesyntax/ColorSupplier.java [hunk 4]
```diff
@@ -0,0 +1,20 @@
...
+
+    public String getRandomColor() {
+        return "" + Colors.values()[getIndex()];
+    }
+
+    public Random getRandom() {
+        return random;
```
Comment (kozhukhovsky) [inline src/main/java/core/basesyntax/ColorSupplier.java:18]:
```suggestion

```

File: src/main/java/core/basesyntax/Colors.java [hunk 5]
```diff
@@ -0,0 +1,9 @@
+package core.basesyntax;
+
+public enum Colors {
```
Comment (kozhukhovsky) [inline src/main/java/core/basesyntax/Colors.java:3]:
```suggestion
public enum Color {

```

File: src/main/java/core/basesyntax/Drawable.java [hunk 6]
```diff
@@ -0,0 +1,7 @@
+package core.basesyntax;
+
+public interface Drawable {
+
+    double findArea();
+}
+
```
Comment (kozhukhovsky) [inline src/main/java/core/basesyntax/Drawable.java:7]:
```suggestion
public interface Drawable {
    double findArea();
}
```

File: src/main/java/core/basesyntax/Figure.java [hunk 7]
```diff
@@ -0,0 +1,20 @@
+package core.basesyntax;
+
+public abstract class Figure implements Drawable {
```
Comment (kozhukhovsky) [inline src/main/java/core/basesyntax/Figure.java:3]:
```suggestion
public abstract class Figure implements Drawable, AreaCalculator {

```

Comment (kozhukhovsky) [reply src/main/java/core/basesyntax/Figure.java:3]: create separate interface for method "findArea" like draw

File: src/main/java/core/basesyntax/Main.java [hunk 8]
```diff
@@ -0,0 +1,17 @@
...
+public class Main {
+    public static void main(String[] args) {
+
+        Figure [] figures = new Figure[20];
+        for (int i = 0; (i < figures.length / 2); i++) {
+            figures[i] = new FigureSupplier().getRandomFigure();
+        }
+        for (int i = (figures.length / 2); i < figures.length; i++) {
+            figures[i] = new FigureSupplier().getDefaultFigure();
+        }
```
Comment (kozhukhovsky) [inline src/main/java/core/basesyntax/Main.java:12]:
```suggestion
        for (int i = 0; (i < figures.length / 2); i++) {
            figures[i] = new FigureSupplier().getRandomFigure();
        }
        for (int i = (figures.length / 2); i < figures.length; i++) {
            figures[i] = new FigureSupplier().getDefaultFigure();
        }

```

use one loop

File: src/main/java/core/basesyntax/Rectangle.java [hunk 9]
```diff
@@ -0,0 +1,33 @@
+package core.basesyntax;
+
+public class Rectangle extends Figure {
+
```
Comment (kozhukhovsky) [inline src/main/java/core/basesyntax/Rectangle.java:4]:
```suggestion
```

File: src/main/java/core/basesyntax/Square.java [hunk 10]
```diff
@@ -0,0 +1,26 @@
+package core.basesyntax;
+
+public class Square extends Figure {
+
```
Comment (kozhukhovsky) [inline src/main/java/core/basesyntax/Square.java:4]:
```suggestion

```

File: src/main/java/core/basesyntax/FigureSupplier.java [hunk 11]
```diff
@@ -0,0 +1,33 @@
...
+                getRandom().nextInt(10));
+        getArray() [3] = new IsoscelesTrapezoid(new ColorSupplier().getRandomColor(),
+                getRandom().nextInt(10), random.nextInt(10),random.nextInt(10));
+        getArray() [4] = new RightTriangle(new ColorSupplier().getRandomColor(),
+                getRandom().nextInt(10), getRandom().nextInt(10));
+        return getArray()[getRandom().nextInt(5)];
+    }
+
+    public Figure getDefaultFigure() {
+        return new Circle("WHITE",10);
```
Comment (kozhukhovsky) [inline src/main/java/core/basesyntax/FigureSupplier.java:31]: move default color and 10 to the class constant

File: src/main/java/core/basesyntax/FigureSupplier.java [hunk 12]
```diff
@@ -0,0 +1,33 @@
...
+    private Random random = new Random();
+    private Figure[]array = new Figure[5];
+
+    public Random getRandom() {
+        return random;
+    }
+
+    public Figure[] getArray() {
+        return array;
+    }
```
Comment (kozhukhovsky) [inline src/main/java/core/basesyntax/FigureSupplier.java:15]:
```suggestion
```

File: src/main/java/core/basesyntax/FigureSupplier.java [hunk 13]
```diff
@@ -0,0 +1,33 @@
+package core.basesyntax;
+
+import java.util.Random;
+
+public class FigureSupplier {
+    private Random random = new Random();
+    private Figure[]array = new Figure[5];
```
Comment (kozhukhovsky) [inline src/main/java/core/basesyntax/FigureSupplier.java:7]:
```suggestion

```

File: src/main/java/core/basesyntax/FigureSupplier.java [hunk 14]
```diff
@@ -0,0 +1,33 @@
+package core.basesyntax;
+
+import java.util.Random;
+
+public class FigureSupplier {
+    private Random random = new Random();
```
Comment (kozhukhovsky) [inline src/main/java/core/basesyntax/FigureSupplier.java:6]: add ColorSupplier variable like random

File: src/main/java/core/basesyntax/FigureSupplier.java [hunk 15]
```diff
@@ -0,0 +1,33 @@
...
+    public Random getRandom() {
+        return random;
+    }
+
+    public Figure[] getArray() {
+        return array;
+    }
+
+    public Figure getRandomFigure() {
+        getArray() [0] = new Circle(new ColorSupplier().getRandomColor(), getRandom().nextInt(10));
```
Comment (kozhukhovsky) [inline src/main/java/core/basesyntax/FigureSupplier.java:18]: Use random to get a random number within the number of shape types.
Then make a switch case, where for each number create a new type of figure.

**Review by kozhukhovsky — CHANGES_REQUESTED** [review]

File: src/main/java/core/basesyntax/ColorSupplier.java [hunk 16]
```diff
@@ -0,0 +1,13 @@
...
+import java.util.Random;
+
+public class ColorSupplier {
+    private Random random = new Random();
+
+    public Color getRandomColor() {
+        int index = random.nextInt(Color.values().length);
+        return Color.values()[index];
+    }
+
```
Comment (okuzan) [inline src/main/java/core/basesyntax/ColorSupplier.java:12]:
```suggestion
```

File: src/main/java/core/basesyntax/Figure.java [hunk 17]
```diff
@@ -0,0 +1,19 @@
+package core.basesyntax;
+
+public abstract class Figure implements CalculateArea, Drawable {
+
```
Comment (okuzan) [inline src/main/java/core/basesyntax/Figure.java:4]:
```suggestion
```

File: src/main/java/core/basesyntax/Figure.java [hunk 18]
```diff
@@ -0,0 +1,19 @@
...
+    public void setColor(Color color) {
+        this.color = color;
+    }
+
+    public Color getColor() {
+        return color;
+    }
+}
+
```
Comment (okuzan) [inline src/main/java/core/basesyntax/Figure.java:19]:
```suggestion
```

File: src/main/java/core/basesyntax/IsoscelesTrapezoid.java [hunk 19]
```diff
@@ -0,0 +1,40 @@
+package core.basesyntax;
+
+public class IsoscelesTrapezoid extends Figure {
+    private int bottom1;
+    private int bottom2;
```
Comment (okuzan) [inline src/main/java/core/basesyntax/IsoscelesTrapezoid.java:5]: don't use numbers in naming

File: src/main/java/core/basesyntax/IsoscelesTrapezoid.java [hunk 20]
```diff
@@ -0,0 +1,40 @@
...
+        return (double) (getBottom1() + getBottom2()) / 2 * getHeight();
+    }
+
+    @Override
+    public void draw() {
+        System.out.println(("Figure: isoscelesTrapezoid, area: " + findArea()
+                + " sq.units, bottom 1: " + getBottom1() + " sq.units, bottom 2: "
+                + getBottom2() + " sq.units, height: " + getHeight()
+                + " units, color: " + getColor()));
+
```
Comment (okuzan) [inline src/main/java/core/basesyntax/IsoscelesTrapezoid.java:38]:
```suggestion
```

File: src/main/java/core/basesyntax/Main.java [hunk 21]
```diff
@@ -0,0 +1,15 @@
+package core.basesyntax;
+
+public class Main {
+    public static void main(String[] args) {
+
+        Figure [] figures = new Figure[20];
```
Comment (okuzan) [inline src/main/java/core/basesyntax/Main.java:6]:
```suggestion
        Figure[] figures = new Figure[20];
```

File: src/main/java/core/basesyntax/Main.java [hunk 22]
```diff
@@ -0,0 +1,15 @@
...
+    public static void main(String[] args) {
+
+        Figure [] figures = new Figure[20];
+        for (int i = 0; (i < figures.length); i++) {
+            figures[i] = (i < (figures.length) / 2 ? new FigureSupplier().getRandomFigure()
+                    : new FigureSupplier().getDefaultFigure());
+        }
+        for (Figure figure : figures) {
+            figure.draw();
+        }
```
Comment (okuzan) [inline src/main/java/core/basesyntax/Main.java:13]:
```suggestion
             figures[i].draw();
        }
```

File: src/main/java/core/basesyntax/Rectangle.java [hunk 23]
```diff
@@ -0,0 +1,32 @@
+package core.basesyntax;
+
+public class Rectangle extends Figure {
+    private int side1;
+    private int side2;
+
```
Comment (okuzan) [inline src/main/java/core/basesyntax/Rectangle.java:6]: same

File: src/main/java/core/basesyntax/RightTriangle.java [hunk 24]
```diff
@@ -0,0 +1,32 @@
+package core.basesyntax;
+
+public class RightTriangle extends Figure {
+    private int side1;
+    private int side2;
```
Comment (okuzan) [inline src/main/java/core/basesyntax/RightTriangle.java:5]: same

**Review by okuzan — CHANGES_REQUESTED** [review]

File: src/main/java/core/basesyntax/FigureSupplier.java [hunk 25]
```diff
@@ -0,0 +1,33 @@
...
+                        random.nextInt(10), random.nextInt(10));
+            case 2:
+                return new Square(colorSupplier.getRandomColor(),
+                        random.nextInt(10));
+            case 3:
+                return new IsoscelesTrapezoid(colorSupplier.getRandomColor(),
+                        random.nextInt(10), random.nextInt(10),random.nextInt(10));
+            default:
+                return new RightTriangle(colorSupplier.getRandomColor(),
+                        random.nextInt(10), random.nextInt(10));
```
Comment (olekskov) [inline src/main/java/core/basesyntax/FigureSupplier.java:26]: magic numbers should be extracted to constants with proper naming

**Review by olekskov — APPROVED** [review]
Good job overall, approved.
Please pay attention to comment below for further tasks

File: src/main/java/core/basesyntax/FigureSupplier.java (hunk 25 above)
Comment (VadimKadasiev) [reply src/main/java/core/basesyntax/FigureSupplier.java:26]: OK. Thank you!

**Pull request closed without merging**